pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -3264,55 +3264,71 @@ protected MessageId internalTerminate(boolean authoritative) {
     }
 
     protected void internalTerminatePartitionedTopic(AsyncResponse asyncResponse, boolean authoritative) {
+        CompletableFuture<Void> future;
         if (topicName.isGlobal()) {
-            validateGlobalNamespaceOwnership(namespaceName);
+            future = validateGlobalNamespaceOwnershipAsync(namespaceName);
+        } else {
+            future = CompletableFuture.completedFuture(null);
         }
-        validateTopicOwnership(topicName, authoritative);
-        validateTopicOperation(topicName, TopicOperation.TERMINATE);
-
-      Map<Integer, MessageId> messageIds = new ConcurrentHashMap<>();
 
-      PartitionedTopicMetadata partitionMetadata = getPartitionedTopicMetadata(topicName, authoritative, false);
-
-        if (partitionMetadata.partitions == 0) {
-            throw new RestException(Status.METHOD_NOT_ALLOWED, "Termination of a non-partitioned topic is "
-                    + "not allowed using partitioned-terminate, please use terminate commands.");
-        }
-        if (partitionMetadata.partitions > 0) {
-          final List<CompletableFuture<MessageId>> futures = Lists.newArrayList();
+        future.thenAccept(__ -> validateTopicOperationAsync(topicName, TopicOperation.TERMINATE)
+                .thenCompose(unused -> getPartitionedTopicMetadataAsync(topicName, authoritative, false))
+                .thenAccept(partitionMetadata -> {
+                    if (partitionMetadata.partitions == 0) {
+                        String msg = "Termination of a non-partitioned topic is not allowed using partitioned-terminate"
+                                + ", please use terminate commands";
+                        log.error("[{}] [{}] {}", clientAppId(), topicName, msg);
+                        asyncResponse.resume(new RestException(Status.METHOD_NOT_ALLOWED, msg));
+                        return;
+                    }
+                    if (partitionMetadata.partitions > 0) {
+                        validateTopicOwnershipAsync(topicName, authoritative).thenAccept(unused -> {
```

Comment:
1. exception of `validateTopicOwnershipAsync` seems not handled.
2. I think we don't need to add `validateTopicOwnershipAsync` here.